### ISSUE TITLE
Warnings for Android NDK build

### ISSUE THREAD

**Issue by noloader** [libexpat/libexpat#383]
Hi Everyone,

I'm working on Android and building [Unbound](), which depends on Expat. I'm using the 2.2.9 release tarball of Expat. I'm also using Android NDK-r20b.

The latest Android NDK's use Clang, not GCC. It is causing some extra noise during the compile of Expat.

```
libtool: compile:  armv7a-linux-androideabi21-clang -DHAVE_CONFIG_H -I. -I.. -DHAVE_EXPAT_CONFIG_H -D__ANDROID_API__=21 -march=armv7-a -mthumb -mfloat-abi=softfp -funwind-tables -fexceptions -frtti --sysroot=/home/jwalton/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot -Wall -Wextra -fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes -pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion -Wshadow -Wformat=2 -Wmisleading-indentation -fvisibility=hidden -DXML_ENABLE_VISIBILITY=1 -MT xmlparse.lo -MD -MP -MF .deps/xmlparse.Tpo -c xmlparse.c  -fPIC -DPIC -o .libs/xmlparse.o
warning: unknown warning option '-Wduplicated-cond'; did you mean
      '-Wduplicate-enum'? [-Wunknown-warning-option]
warning: unknown warning option '-Wduplicated-branches'
      [-Wunknown-warning-option]
warning: unknown warning option '-Wlogical-op'; did you mean '-Wlong-long'?
      [-Wunknown-warning-option]
warning: unknown warning option '-Wrestrict' [-Wunknown-warning-option]
warning: unknown warning option '-Wjump-misses-init' [-Wunknown-warning-option]
warning: unknown warning option '-Wmisleading-indentation'; did you mean
      '-Wbinding-in-condition'? [-Wunknown-warning-option]
6 warnings generated.
```

The six warnings are repeated often during the build.

**Comment by hartwork** [comment]
`configure.ac` is adding those flags as of today:
```console
# for i in -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wrestrict -Wjump-misses-init -Wmisleading-indentation ; do git --no-pager grep -- $i ; done | sort -u
configure.ac:   AX_APPEND_COMPILE_FLAGS([-pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op], [CFLAGS])
configure.ac:   AX_APPEND_COMPILE_FLAGS([-Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion], [CFLAGS])
configure.ac:   AX_APPEND_COMPILE_FLAGS([-Wshadow -Wformat=2 -Wmisleading-indentation], [CFLAGS])])
```
The Autotools build system is likely to be dropped in favor of CMake soon (#358). If it was errors it would be a different story. I'm not sure if it's worth your or my time fixing these warnings. What do you think?

**Comment by noloader** [comment]
Thanks @hartwork.

The odd thing (to me) is [the guard arounf line 70](https://github.com/libexpat/libexpat/blob/master/expat/configure.ac#L70):

```
AS_IF([test "$GCC" = yes],
  [AX_APPEND_COMPILE_FLAGS([-Wall -Wextra], [CFLAGS])
   dnl Be careful about adding the -fexceptions option; some versions of
   dnl GCC don't support it and it causes extra warnings that are only
   dnl distracting; avoid.
   AX_APPEND_COMPILE_FLAGS([-fexceptions], [CFLAGS])
   AX_APPEND_COMPILE_FLAGS([-fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes], [CFLAGS])
   AX_APPEND_COMPILE_FLAGS([-pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op], [CFLAGS])
   AX_APPEND_COMPILE_FLAGS([-Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion], [CFLAGS])
   AX_APPEND_COMPILE_FLAGS([-Wshadow -Wformat=2 -Wmisleading-indentation], [CFLAGS])])
```

The [compiler names](https://github.com/noloader/unbound/blob/master/android/setenv_android.sh#L57) are `armv7a-linux-androideabi$ANDROID_API-clang`, `aarch64-linux-android$ANDROID_API-clang`, `i686-linux-android$ANDROID_API-clang`, `x86_64-linux-android$ANDROID_API-clang`, etc (where `ANDROID_API` is a number like 23).

How are they matching GCC?

**Comment by noloader** [comment]
Thanks again @hartwork.

> The Autotools build system is likely to be dropped in favor of CMake soon

Do build systems need to be mutually exclusive? Can you select one as the primary, and support both?

The reason I bring it up is, I help maintain a large project. We use GNU Make as the primary build system on Unix and Linux (yep, you can do everything you need with just Gmake). We support Autotools and Cmake as secondary build systems.

Regarding CMake, CMake does not use recommended platform flags. The project will be built with the wrong `CFLAGS` and `CXXFLAGS` out of the box. For example, here are the `CFLAGS` and `CXXFLAGS` that should be used for [Android projects](https://github.com/noloader/unbound/blob/master/android/setenv_android.sh#L57). The flags will be missing if you use CMake to build the project. There are other problems with CMake in practice, but the flags are important to me because they cause subtle, hard to track down problems.

**Comment by hartwork** [comment]
> How are they matching GCC?

`GCC=yes` is set by `configure` as rendered by GNU Autoconf (and Expat started trusting it in 286151fe68183c892d1f30ae4eb68c7e7cb4cbc8). I'm not sure why Android Clang is classified as GCC by them. We can adjust the check in `configure.ac` if you have a good idea how.

> Do build systems need to be mutually exclusive? Can you select one as the primary, and support both?

It's a pain to maintain and to extend multiple build systems over time. Expat has had both Autotools and CMake for years: it's not ideal.

> The reason I bring it up is, I help maintain a large project. We use GNU Make as the primary build system on Unix and Linux (yep, you can do everything you need with just Gmake). We support Autotools and Cmake as secondary build systems.
> 
> Regarding CMake, CMake does not use recommended platform flags. The project will be built with the wrong `CFLAGS` and `CXXFLAGS` out of the box. For example, here are the `CFLAGS` and `CXXFLAGS` that should be used for [Android projects](https://github.com/noloader/unbound/blob/master/android/setenv_android.sh#L57). The flags will be missing if you use CMake to build the project. There are other problems with CMake in practice, but the flags are important to me because they cause subtle, hard to track down problems.

I'm not sure I fully understand what that problem is.

To pass https://github.com/noloader/unbound/blob/master/android/setenv_android.sh#L57 to CMake I would go for something like this:
- `-DCMAKE_BUILD_TYPE=Custom`  # to prevent CMake from adding their own defaults
- `-DCMAKE_C_FLAGS=...`
- `-DCMAKE_CXX_FLAGS=...`
- `-DCMAKE_C_COMPILER=...`
- `-DCMAKE_CXX_COMPILER=...`
- ...

(Or make a new toolchain file for that.)

If any of those do not work for Expat, let's fix Expat's CMake build system.

**Comment by noloader** [comment]
@hartwork,

I was looking at the output of `configure`. The `install_expat.sh` script can be found [here](https://github.com/NLnetLabs/unbound/blob/master/contrib/android/install_expat.sh).

```
Configuring Expat
checking build system type... x86_64-pc-linux-gnu
checking host system type... armv7a-unknown-linux-androideabi
checking for a BSD-compatible install... /usr/bin/install -c
...
checking whether C compiler accepts -Wall... yes
checking whether C compiler accepts -Wextra... yes
checking whether C compiler accepts -fexceptions... yes
checking whether C compiler accepts -fno-strict-aliasing... yes
checking whether C compiler accepts -Wmissing-prototypes... yes
checking whether C compiler accepts -Wstrict-prototypes... yes
checking whether C compiler accepts -pedantic... yes
checking whether C compiler accepts -Wduplicated-cond... yes
checking whether C compiler accepts -Wduplicated-branches... yes
checking whether C compiler accepts -Wlogical-op... yes
checking whether C compiler accepts -Wrestrict... yes
checking whether C compiler accepts -Wnull-dereference... yes
checking whether C compiler accepts -Wjump-misses-init... yes
checking whether C compiler accepts -Wdouble-promotion... yes
checking whether C compiler accepts -Wshadow... yes
checking whether C compiler accepts -Wformat=2... yes
checking whether C compiler accepts -Wmisleading-indentation... yes
checking whether we are using the GNU C++ compiler... yes
```

I'm wondering if the correct compiler is being used for the tests. Or maybe the test is broken.

I've seen the Autotools tests say "success" because the compiler returns 0 even though the test fails in practice. It may be happening with the warnings.

The pathological Autotools case I encountered was on Solaris. SunCC says `-xarch=sha` is an illegal option but the compiler returns 0 (success), not 1 (failure). Later, when we try to compile SHA code with the `-xarch=sha` option the compile and link fails. We actually had to write our own `COMPILE_ELSEIF` macro to work around it; see [`XXX_COMPILE_IFELSE`](https://github.com/noloader/cryptopp-autotools/blob/master/configure.ac#L161) and [`XXX_LINK_IFELSE`](https://github.com/noloader/cryptopp-autotools/blob/master/configure.ac#L172).

**Comment by hartwork** [comment]
I can confirm now that the compiler checks enly check for non-zero exit codes, not for warnings and do not use `-Werror`. So this is how ` -Wduplicated-cond` is mis-detected to be surpported by Clang 9, quoting `config.log`:
```
[..]
configure:9846: checking whether C compiler accepts -Wduplicated-cond
configure:9865: clang-9 -c -g -O2 -Wall -Wextra -fexceptions -fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes -pedantic  -Wduplicated-cond -DHAVE_EXPAT_CONFIG_H conftest.c >&5
warning: unknown warning option '-Wduplicated-cond'; did you mean '-Wduplicate-enum'? [-Wunknown-warning-option]
1 warning generated.
configure:9865: $? = 0
[..]
```

Maybe configuring with `./configure CFLAGS="-Wno-unknown-warning-option"` works for a global silencer?

What other options do we have?

**Comment by noloader** [comment]
I think this is part of the solution:

```
$ git diff
diff --git a/expat/configure.ac b/expat/configure.ac
index 96fbe00..b10bc43 100644
--- a/expat/configure.ac
+++ b/expat/configure.ac
@@ -69,6 +69,9 @@ AC_SUBST(LIBAGE)
 
 AC_LANG([C])
 AC_PROG_CC_C99
+
+OLD_CFLAGS="$CFLAGS"
+CFLAGS="-Werror"
 AS_IF([test "$GCC" = yes],
   [AX_APPEND_COMPILE_FLAGS([-Wall -Wextra], [CFLAGS])
    dnl Be careful about adding the -fexceptions option; some versions of
@@ -79,6 +82,7 @@ AS_IF([test "$GCC" = yes],
    AX_APPEND_COMPILE_FLAGS([-pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op], [CFLAGS])
    AX_APPEND_COMPILE_FLAGS([-Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion], [CFLAGS])
    AX_APPEND_COMPILE_FLAGS([-Wshadow -Wformat=2 -Wmisleading-indentation], [CFLAGS])])
+CFLAGS="$OLD_CFLAGS"
 
 AC_LANG_PUSH([C++])
 AC_PROG_CXX
```

I realize the Autotools changes will be blown away `CFLAGS="$OLD_CFLAGS"`.

But stepping back to 10,000 feet... The user owns `CPPFLAGS`, `CFLAGS` and `CXXFLAGS`. Autotools projects should not be messing with them. Autotools projects should use `AM_CPPFLAGS`, `AM_CFLAGS` and `AM_CXXFLAGS`. Using the `AM_*` variables  results in something like this:

```
$ git diff
diff --git a/expat/configure.ac b/expat/configure.ac
index 96fbe00..d75e859 100644
--- a/expat/configure.ac
+++ b/expat/configure.ac
@@ -69,16 +69,20 @@ AC_SUBST(LIBAGE)
 
 AC_LANG([C])
 AC_PROG_CC_C99
+
+OLD_CFLAGS="$CFLAGS"
+CFLAGS="-Werror"
 AS_IF([test "$GCC" = yes],
-  [AX_APPEND_COMPILE_FLAGS([-Wall -Wextra], [CFLAGS])
+  [AX_APPEND_COMPILE_FLAGS([-Wall -Wextra], [AM_CFLAGS])
    dnl Be careful about adding the -fexceptions option; some versions of
    dnl GCC don't support it and it causes extra warnings that are only
    dnl distracting; avoid.
-   AX_APPEND_COMPILE_FLAGS([-fexceptions], [CFLAGS])
-   AX_APPEND_COMPILE_FLAGS([-fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes], [CFLAGS])
-   AX_APPEND_COMPILE_FLAGS([-pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op], [CFLAGS])
-   AX_APPEND_COMPILE_FLAGS([-Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion], [CFLAGS])
-   AX_APPEND_COMPILE_FLAGS([-Wshadow -Wformat=2 -Wmisleading-indentation], [CFLAGS])])
+   AX_APPEND_COMPILE_FLAGS([-fexceptions], [AM_CFLAGS])
+   AX_APPEND_COMPILE_FLAGS([-fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes], [AM_CFLAGS])
+   AX_APPEND_COMPILE_FLAGS([-pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op], [AM_CFLAGS])
+   AX_APPEND_COMPILE_FLAGS([-Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion], [AM_CFLAGS])
+   AX_APPEND_COMPILE_FLAGS([-Wshadow -Wformat=2 -Wmisleading-indentation], [AM_CFLAGS])])
+CFLAGS="$OLD_CFLAGS"
 
 AC_LANG_PUSH([C++])
 AC_PROG_CXX
```

Then, at the bottom `configure.ac`:

```
## Write the Automake flags we set
AC_SUBST([AM_CPPFLAGS])
AC_SUBST([AM_CFLAGS])
AC_SUBST([AM_CXXFLAGS])
AC_SUBST([AM_LDFLAGS])
```

When Autools produces the Makefile from the `Makefile.am`, a recipe will look something like this (with some hand waiving):

```
*.o : *.c:
    $(CC) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) ...
*.o : *.cpp:
    $(CXX) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) ...
```

The project's `AM_CPPFLAGS`, `AM_CPPFLAGS`, `AM_CXXFLAGS` will be present, but the user's `CPPFLAGS`, `CPPFLAGS`, `CXXFLAGS` flags will trump the projects flags (like they should). The user always has the final say in flags.

Many projects miss this detail. And it is not obvious things need to be wired like that from the [GNU Coding Standards](https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html#Makefile-Conventions). You basically have to take what the manual tells you about "the user owns CFLAGS; and the user's CFLAGS trump the project's flags", and turn them into an implementation using `CFLAGS` and `AM_CFLAGS` (and friends).

I can try to work up a PR if you are interested.

**Comment by noloader** [comment]
Switching to `AM_CPPFLAGS`, `AM_CFLAGS` and `AM_CXXFLAGS` looks good. After running `buildconf.sh` and then `configure`:

```
$ cat Makefile | grep -E 'AM_CPPFLAGS|AM_CFLAGS|AM_CXXFLAGS' | fold -w80
AM_CFLAGS = -Wall -Wextra -fexceptions -fno-strict-aliasing -Wmissing-prototypes
 -pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wrestrict -Wnul
l-dereference -Wjump-misses-init -Wdouble-promotion -Wshadow -Wformat=2 -Wmislea
ding-indentation -fvisibility=hidden
AM_CPPFLAGS = -DHAVE_EXPAT_CONFIG_H -DXML_ENABLE_VISIBILITY=1
AM_CXXFLAGS = -Wall -Wextra -fexceptions -fno-strict-aliasing
```

And then running make:

```
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..  -D
HAVE_EXPAT_CONFIG_H -DXML_ENABLE_VISIBILITY=1  -Wall -Wextra -fexceptions -fno-s
trict-aliasing -Wmissing-prototypes -pedantic -Wduplicated-cond -Wduplicated-bra
nches -Wlogical-op -Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-pro
motion -Wshadow -Wformat=2 -Wmisleading-indentation -fvisibility=hidden -g -O2 -
MT xmlparse.lo -MD -MP -MF .deps/xmlparse.Tpo -c -o xmlparse.lo xmlparse.c
```

Let me offer a PR for you to look at.

**Comment by noloader** [comment]
With the PR one can now perform:

```
$ CFLAGS="-DNDEBUG -g2 -O3 -march=native" CXXFLAGS="-DNDEBUG -g2 -O3 -march=native" ./configure
```

Configure will summarize as:

```
Automake flags (can be overridden by user flags):
AM_CPPFLAGS: -DHAVE_EXPAT_CONFIG_H -DXML_ENABLE_VISIBILITY=1
AM_CFLAGS: -Wall -Wextra -fexceptions -fno-strict-aliasing -Wmissing-prototypes -pedantic -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-promotion -Wshadow -Wformat=2 -Wmisleading-indentation -fvisibility=hidden
AM_CXXFLAGS: -Wall -Wextra -fexceptions -fno-strict-aliasing
AM_LDFLAGS: -fno-strict-aliasing

User flags (overrides Automake flags on conflict):
CPPFLAGS: 
CFLAGS: -DNDEBUG -g2 -O3 -march=native
CXXFLAGS: -DNDEBUG -g2 -O3 -march=native
LDFLAGS: 
```

And a make results in the following. Notice the user's flags come last, so the user's flags trump the project's flags (like they should).

```
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..  -D
HAVE_EXPAT_CONFIG_H -DXML_ENABLE_VISIBILITY=1  -Wall -Wextra -fexceptions -fno-s
trict-aliasing -Wmissing-prototypes -pedantic -Wduplicated-cond -Wduplicated-bra
nches -Wlogical-op -Wrestrict -Wnull-dereference -Wjump-misses-init -Wdouble-pro
motion -Wshadow -Wformat=2 -Wmisleading-indentation -fvisibility=hidden -DNDEBUG
 -g2 -O3 -march=native -MT xmlparse.lo -MD -MP -MF .deps/xmlparse.Tpo -c -o xmlp
arse.lo xmlparse.c
```

Good to go :)

**Comment by noloader** [comment]
You can see the results of PR and PR at [Noloader | Expat](https://travis-ci.org/github/noloader/libexpat/builds/662128174). Notice the unknown warnings have disappeared.

There are some new warnings, though:

```
armv7a-linux-androideabi26-clang -DHAVE_CONFIG_H -I. -I..  -I./../lib  -D__ANDRO
ID_API__=26 -Wall -Wextra -fexceptions -fno-strict-aliasing -Wmissing-prototypes
 -Wstrict-prototypes -pedantic -Wnull-dereference -Wdouble-promotion -Wshadow -W
format=2 -fvisibility=hidden -march=armv7-a -mthumb -mfloat-abi=softfp -funwind-
tables -fexceptions --sysroot=/home/travis/ndk/toolchains/llvm/prebuilt/linux-x8
6_64/sysroot -MT xmlwf-codepage.o -MD -MP -MF .deps/xmlwf-codepage.Tpo -c -o xml
wf-codepage.o `test -f codepage.c || echo ./`codepage.c`

xmlfile.c:194:9: warning: implicit declaration of function 'close' is invalid in
      C99 [-Wimplicit-function-declaration]
        close(fd);
        ^
xmlfile.c:194:9: warning: this function declaration is not a prototype
      [-Wstrict-prototypes]
xmlfile.c:199:13: warning: mv -f .deps/xmlwf-codepage.Tpo .deps/xmlwf-codepage.Po
implicit declaration of function 'read' is invalid in
      C99 [-Wimplicit-function-declaration]
    nread = read(fd, buf, READ_SIZE);
            ^
xmlfile.c:199:13: warning: this function declaration is not a prototype
      [-Wstrict-prototypes]
xmlfile.c:203:9: warning: implicit declaration of function 'close' is invalid in
      C99 [-Wimplicit-function-declaration]
        close(fd);
        ^
xmlfile.c:209:9: warning: implicit declaration of function 'close' is invalid in
      C99 [-Wimplicit-function-declaration]
        close(fd);
        ^
xmlfile.c:214:9: warning: implicit declaration of function 'close' is invalid in
      C99 [-Wimplicit-function-declaration]
        close(fd);
        ^
```

**Comment by hartwork** [comment]
I have commented on #387 now.

Regarding `unistd.h`: Windows doesn't have it so it needs to be a conditional include. We happen to already have one a few lines down at https://github.com/libexpat/libexpat/blob/f1ac70b3e31e703678c58e0d54b4e2c37b02eb97/expat/xmlwf/xmlfile.c#L55-L57

So if the include doesn't work I guess something about `HAVE_UNISTD_H` is not working? Please dedicate a new ticket and/or pull request to that — it's quite different to the other things in here.

**Comment by hartwork** [comment]
> There are some new warnings, though. Posix says to [`#include <unistd.h>`](https://pubs.opengroup.org/onlinepubs/009695399/functions/close.html) for the declarations.

FYI It turns out that's a regression resulting from the move to `AM_CPPFLAGS` that makes use lose `-DHAVE_EXPAT_CONFIG_H` for `xmlfile.c`. I have a fix in #393, commit 38ee338b63d07f08a1ac55d0ad592bcf2940a662 precisely.

**Comment by hartwork** [comment]
Fixed by #393 , closing.
